### PR TITLE
Client Encryption : Adds integration with latest Cryptography package and removes caching of AeadAes256CbcHmac256EncryptionAlgorithm object

### PR DESCRIPTION
**Pull request by kr-santosh**
## Description

1. This PR removes the caching of AeadAes256CbcHmac256EncryptionAlgorithm object which was done earlier and now completely relies on MDE layer for caching. 
2. Upgrades to latest "Microsoft.Data.Encryption.Cryptography" Version="0.2.0-pre".

## Type of change

Please delete options that are not relevant.

- [] Bug fix (non-breaking change which fixes an issue)
- [] This change requires a documentation update

### CODE REVIEW

File: Microsoft.Azure.Cosmos.Encryption/src/EncryptionCosmosClientExtensions.cs [hunk 1]
```diff
@@ -32,6 +32,9 @@ public static class EncryptionCosmosClientExtensions
                 throw new ArgumentNullException(nameof(cosmosClient));
             }
 
+            // set the TTL for ProtectedDataEncryption at the Encryption CosmosClient Init so that we have a uniform expiry of the KeyStoreProvider and ProtectedDataEncryption cache items.
+            ProtectedDataEncryptionKey.TimeToLive = (TimeSpan)encryptionKeyStoreProvider.DataEncryptionKeyCacheTimeToLive;
```
Comment (abhijitpai) [inline Microsoft.Azure.Cosmos.Encryption/src/EncryptionCosmosClientExtensions.cs:36]: It looks like the underlying intent is to forever cache when the TTL is explicitly set to null on the encryptionKeyStoreProvider. So we should do:

if(encryptionKeyStoreProvider.DataEncryptionKeyCacheTimeToLive.HasValue)
{
ProtectedDataEncryptionKey.TimeToLive = encryptionKeyStoreProvider.DataEncryptionKeyCacheTimeToLive.Value;
}
else
{
ProtectedDataEncryptionKey.TimeToLive = TimeSpan.MaxValue;
}

File: Microsoft.Azure.Cosmos.Encryption/src/Custom/CosmosDataEncryptionKeyProvider.cs [hunk 2]
```diff
@@ -86,14 +88,21 @@ internal Container Container
             this.dataEncryptionKeyContainerCore = new DataEncryptionKeyContainerCore(this);
             this.DekCache = new DekCache(dekPropertiesTimeToLive);
             this.PdekCacheTimeToLive = cacheTimeToLive;
+            if (this.PdekCacheTimeToLive.HasValue)
+            {
+                // set the TTL for Protected Data Encryption.
+                ProtectedDataEncryptionKey.TimeToLive = (TimeSpan)this.PdekCacheTimeToLive;
```
Comment (abhijitpai) [inline Microsoft.Azure.Cosmos.Encryption/src/Custom/CosmosDataEncryptionKeyProvider.cs:94]: Use .Value instead of cast.

**Comment by abhijitpai** [comment]
>                 false);

parameterName:

---
Refers to: Microsoft.Azure.Cosmos.Encryption/src/EncryptionSettings.cs:53 in 86ed4b5. [](commit_id = 86ed4b57b9bbcb36421b0b4cf35a4a3a571ca631, deletion_comment = False)

File: Microsoft.Azure.Cosmos.Encryption/src/EncryptionSettings.cs [hunk 3]
```diff
@@ -80,76 +58,90 @@ public EncryptionSettings()
...
-                            case CosmosEncryptionType.Plaintext:
-                                encryptionType = EncryptionType.Plaintext;
-                                break;
-                            default:
-                                throw new ArgumentException($"Invalid encryption type {propertyToEncrypt.EncryptionType}. Please refer to https://aka.ms/CosmosClientEncryption for more details. ");
-                        }
-
-                        encryptionSettings = EncryptionSettings.Create(encryptionSettings, encryptionType);
-                        return new CachedEncryptionSettings(encryptionSettings, encryptionSettings.EncryptionSettingTimeToLive);
+                        return EncryptionSettings.Create(encryptionSettings);
```
Comment (abhijitpai) [inline Microsoft.Azure.Cosmos.Encryption/src/EncryptionSettings.cs:69]: >n Encry [](start = 29, length = 7)

It looks like we have made this unnecessarily confusing by having this class represent both a setting and a map of settings by column name. Can we use two different classes?

File: Microsoft.Azure.Cosmos.Encryption/src/QueryDefinitionExtensions.cs [hunk 4]
```diff
@@ -99,14 +99,18 @@ public static class QueryDefinitionExtensions
                     throw new ArgumentException($"Unsupported argument with Path: {path} for query. For executing queries on encrypted path requires the use of deterministic encryption type. Please refer to https://aka.ms/CosmosClientEncryption for more details. ");
                 }
 
+                AeadAes256CbcHmac256EncryptionAlgorithm aeadAes256CbcHmac256EncryptionAlgorithm = await settings.BuildEncryptionAlgorithmForSettingAsync(
+                    encryptionContainer.EncryptionProcessor,
+                    cancellationToken: default);
```
Comment (abhijitpai) [inline Microsoft.Azure.Cosmos.Encryption/src/QueryDefinitionExtensions.cs:104]: Pass the cancellationToken that was passed in

**Comment by abhijitpai** [comment]
>             this.EncryptionSettings = new EncryptionSettings();

If this settings map is a child of the processor, just pass it the processor here rather than passing it in every method.

---
Refers to: Microsoft.Azure.Cosmos.Encryption/src/EncryptionProcessor.cs:54 in 86ed4b5. [](commit_id = 86ed4b57b9bbcb36421b0b4cf35a4a3a571ca631, deletion_comment = False)

File: Microsoft.Azure.Cosmos.Encryption/src/EncryptionCosmosClientExtensions.cs (hunk 1 above)
Comment (kr-santosh) [reply Microsoft.Azure.Cosmos.Encryption/src/EncryptionCosmosClientExtensions.cs]: Fixed. Setting to 100 years.

File: Microsoft.Azure.Cosmos.Encryption/src/EncryptionSettings.cs (hunk 3 above)
Comment (kr-santosh) [reply Microsoft.Azure.Cosmos.Encryption/src/EncryptionSettings.cs]: Fixed.

File: Microsoft.Azure.Cosmos.Encryption/src/QueryDefinitionExtensions.cs (hunk 4 above)
Comment (kr-santosh) [reply Microsoft.Azure.Cosmos.Encryption/src/QueryDefinitionExtensions.cs]: Fixed.

File: Microsoft.Azure.Cosmos.Encryption/src/Custom/CosmosDataEncryptionKeyProvider.cs (hunk 2 above)
Comment (kr-santosh) [reply Microsoft.Azure.Cosmos.Encryption/src/Custom/CosmosDataEncryptionKeyProvider.cs]: Fixed.

**Review by abhijitpai — DISMISSED** [review]
:shipit:

File: Microsoft.Azure.Cosmos.Encryption/src/Custom/CosmosDataEncryptionKeyProvider.cs [hunk 5]
```diff
@@ -74,39 +74,55 @@ internal Container Container
...
+            this.PdekCacheTimeToLive = this.EncryptionKeyStoreProvider.DataEncryptionKeyCacheTimeToLive;
+            if (this.PdekCacheTimeToLive.HasValue)
+            {
+                // set the TTL for Protected Data Encryption.
+                ProtectedDataEncryptionKey.TimeToLive = this.PdekCacheTimeToLive.Value;
+            }
+            else
+            {
+                // arbitrarily large caching period.
+                ProtectedDataEncryptionKey.TimeToLive = TimeSpan.FromDays(36500);
```
Comment (anujtoshniwal) [inline Microsoft.Azure.Cosmos.Encryption/src/Custom/CosmosDataEncryptionKeyProvider.cs:95]: Should we not stick to 2hrs default here? Why use an arbitrarily large caching period?

Comment (kr-santosh) [reply Microsoft.Azure.Cosmos.Encryption/src/Custom/CosmosDataEncryptionKeyProvider.cs:95]: we will follow the similar model as public and have Pdek and StoreProvider Cache TTL align, since setting DataEncryptionKeyCacheTimeToLive to null would result in forever caching (so getting rid of cacheTimeToLive). ProtectedDataEncryptionKey cache is not able to handle Timespan.MaxValue hence that value.

Comment (anujtoshniwal) [reply Microsoft.Azure.Cosmos.Encryption/src/Custom/CosmosDataEncryptionKeyProvider.cs:95]: Can you please add the reason to the comment?

File: Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs [hunk 6]
```diff
@@ -38,7 +38,10 @@ public class MdeEncryptionTests
         public static async Task ClassInitialize(TestContext context)
         {
             MdeEncryptionTests.client = TestCommon.CreateCosmosClient();
-            testEncryptionKeyStoreProvider = new TestEncryptionKeyStoreProvider();
+            testEncryptionKeyStoreProvider = new TestEncryptionKeyStoreProvider
+            {
+                DataEncryptionKeyCacheTimeToLive = null
```
Comment (anujtoshniwal) [inline Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs:43]: Can you also please add a test with default value of DataEncryptionKeyCacheTimeToLive? User will most likely not set it to null..

Comment (kr-santosh) [reply Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs:43]: Fixed.

File: Microsoft.Azure.Cosmos.Encryption/src/Custom/CosmosDataEncryptionKeyProvider.cs (hunk 5 above)
Comment (kr-santosh) [reply Microsoft.Azure.Cosmos.Encryption/src/Custom/CosmosDataEncryptionKeyProvider.cs:97]: Fixed.

File: Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs [hunk 7]
```diff
@@ -823,14 +826,26 @@ public async Task EncryptionRestrictedProperties()
...
         {
+            CosmosClient clientWithNoCaching = TestCommon.CreateCosmosClient(builder => builder
+                .Build());
+
+            TestEncryptionKeyStoreProvider testEncryptionKeyStoreProvider = new TestEncryptionKeyStoreProvider
+            {
+                DataEncryptionKeyCacheTimeToLive = TimeSpan.Zero
+            };
+
+            CosmosClient encryptionCosmosClientWithBulk = clientWithNoCaching.WithEncryption(testEncryptionKeyStoreProvider);
```
Comment (anujtoshniwal) [inline Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs:837]: rename

File: Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs [hunk 8]
```diff
@@ -1041,12 +1058,38 @@ public async Task EncryptionTransactionalBatchWithCustomSerializer()
...
-                await MdeEncryptionTests.MdeCreateItemAsync(MdeEncryptionTests.encryptionContainer);
+            // Default cache TTL 2 hours.
+            TestEncryptionKeyStoreProvider newtestEncryptionKeyStoreProvider = new TestEncryptionKeyStoreProvider();           
+            CosmosClient newEncryptionClient = MdeEncryptionTests.client.WithEncryption(newtestEncryptionKeyStoreProvider);
+            Database database = newEncryptionClient.GetDatabase(MdeEncryptionTests.database.Id);
+
+            Container encryptionContainer = database.GetContainer(MdeEncryptionTests.encryptionContainer.Id);
+
+            for (int i = 0; i < 2; i++)
+                await MdeEncryptionTests.MdeCreateItemAsync(encryptionContainer);
```
Comment (anujtoshniwal) [inline Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs:1069]: use brackets please

**Review by anujtoshniwal — DISMISSED** [review]

File: Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs (hunk 7 above)
Comment (kr-santosh) [reply Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs]: Fixed.

File: Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs (hunk 8 above)
Comment (kr-santosh) [reply Microsoft.Azure.Cosmos.Encryption/tests/EmulatorTests/MdeEncryptionTests.cs:1070]: Fixed.

**Review by abhijitpai — APPROVED** [review]

**Review by j82w — APPROVED** [review]
